

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Guide &mdash; hairpin2 3.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/custom.css?v=be469e84" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=acc74ff5"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="hairpin2 package" href="hairpin2.html" />
    <link rel="prev" title="Quickstart" href="quickstart.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            hairpin2
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a><ul>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#usage">Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#command-line-interface">Command Line Interface</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#hp2-utils">hp2-utils</a><ul>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#explain-var">explain-var</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#get-params">get-params</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#rationale-assumptions-limitations">Rationale, Assumptions &amp; Limitations</a></li>
<li class="toctree-l2"><a class="reference internal" href="#processes-and-parameters">Processes and Parameters</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#process-relationships-reference">Process Relationships Reference</a></li>
<li class="toctree-l3"><a class="reference internal" href="#read-taggers">Read Taggers</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#mark-support">mark-support</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mark-overlap">mark-overlap</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mark-low-qual">mark-low-qual</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mark-duplicates">mark-duplicates</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#variant-flaggers">Variant Flaggers</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#lqf">LQF</a></li>
<li class="toctree-l4"><a class="reference internal" href="#dvf">DVF</a></li>
<li class="toctree-l4"><a class="reference internal" href="#alf">ALF</a></li>
<li class="toctree-l4"><a class="reference internal" href="#adf">ADF</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#writing-a-config">Writing a Config</a></li>
<li class="toctree-l2"><a class="reference internal" href="#understanding-decisions">Understanding Decisions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reproducibility-parameter-distribution">Reproducibility &amp; Parameter Distribution</a></li>
<li class="toctree-l2"><a class="reference internal" href="#further-development">Further Development</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="hairpin2.html">hairpin2 package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="hairpin2.html#scientific-modules">Scientific Modules</a><ul>
<li class="toctree-l3"><a class="reference internal" href="hairpin2.const.html">const module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.const.html#hairpin2.const.ValidNucleotides"><code class="docutils literal notranslate"><span class="pre">ValidNucleotides</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Strand"><code class="docutils literal notranslate"><span class="pre">Strand</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.const.html#hairpin2.const.MutTypes"><code class="docutils literal notranslate"><span class="pre">MutTypes</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags"><code class="docutils literal notranslate"><span class="pre">Tags</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.const.html#hairpin2.const.TaggerNamespaces"><code class="docutils literal notranslate"><span class="pre">TaggerNamespaces</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.const.html#hairpin2.const.FlaggerNamespaces"><code class="docutils literal notranslate"><span class="pre">FlaggerNamespaces</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="hairpin2.sci_funcs.html">sci_funcs module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.sci_funcs.html#hairpin2.sci_funcs.TagSupportingReads"><code class="docutils literal notranslate"><span class="pre">TagSupportingReads</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.sci_funcs.html#hairpin2.sci_funcs.TagLowQualReads"><code class="docutils literal notranslate"><span class="pre">TagLowQualReads</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.sci_funcs.html#hairpin2.sci_funcs.TagFragmentReads"><code class="docutils literal notranslate"><span class="pre">TagFragmentReads</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.sci_funcs.html#hairpin2.sci_funcs.TagStutterDuplicateReads"><code class="docutils literal notranslate"><span class="pre">TagStutterDuplicateReads</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.sci_funcs.html#hairpin2.sci_funcs.AlignmentScoreTest"><code class="docutils literal notranslate"><span class="pre">AlignmentScoreTest</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.sci_funcs.html#hairpin2.sci_funcs.AnomalousDistributionTest"><code class="docutils literal notranslate"><span class="pre">AnomalousDistributionTest</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.sci_funcs.html#hairpin2.sci_funcs.ProportionBasedTest"><code class="docutils literal notranslate"><span class="pre">ProportionBasedTest</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="hairpin2.main.html">main module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.main.html#hairpin2.main.hairpin2"><code class="docutils literal notranslate"><span class="pre">hairpin2()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="hairpin2.html#subpackages">Subpackages</a><ul>
<li class="toctree-l3"><a class="reference internal" href="hairpin2.infrastructure.html">infrastructure package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.infrastructure.html#submodules">Submodules</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="hairpin2.process_wrappers.html">process_wrappers package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="hairpin2.process_wrappers.html#submodules">Submodules</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="dev.html">Developer Documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">hairpin2</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content style-external-links">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Guide</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/guide.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="guide">
<span id="guide-target"></span><h1>Guide<a class="headerlink" href="#guide" title="Link to this heading"></a></h1>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>It is recommended to read <a class="reference internal" href="quickstart.html#quickstart-target"><span class="std std-ref">Quickstart</span></a> prior to reading the guide.</p>
</div>
<p>Given a VCF, and alignment files for the relevant samples of that VCF, <code class="docutils literal notranslate"><span class="pre">hairpin2</span></code> will return a VCF with variants flagged in the FILTER column as follows:</p>
<table class="docutils align-default" style="width: 100">
<colgroup>
<col style="width: 15.0%" />
<col style="width: 25.0%" />
<col style="width: 60.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>FILTER</p></th>
<th class="head"><p>Expanded Name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>ADF</strong></p></td>
<td><p><strong>A</strong>nomalous <strong>D</strong>istribution <strong>F</strong>lag</p></td>
<td><p>An implementation and extension of the artefact detection algorithm described in <a class="reference external" href="https://www.nature.com/articles/s41596-020-00437-6">Ellis et al, 2020</a>. It detects variants which appear with anomalously regular positional distribution in supporting reads.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>ALF</strong></p></td>
<td><p><strong>AL</strong>ignment score <strong>F</strong>lag</p></td>
<td><p>Indicates variants which are supported by reads with poor signal-to-noise, per the alignment score. It is complementary to the <code class="docutils literal notranslate"><span class="pre">ADF</span></code> flag – artefacts with anomalous distributions are often accompanied by a marked decrease in alignment score.</p></td>
</tr>
<tr class="row-even"><td><p><strong>DVF</strong></p></td>
<td><p><strong>D</strong>uplication <strong>V</strong>ariant <strong>F</strong>lag</p></td>
<td><p>Uses a relatively naive but effective algorithm for detecting variants that are the result of <a class="reference internal" href="#dup-explain-target">hidden duplicates</a>, themselves a result of PCR error.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>LQF</strong></p></td>
<td><p><strong>L</strong>ow <strong>Q</strong>uality <strong>F</strong>lag</p></td>
<td><p>Tests whether a variant is largely supported by low quality reads.</p></td>
</tr>
</tbody>
</table>
<p>Further details of the implementation of each flag can be found by reading the <a class="reference internal" href="#process-target">Process and Parameters</a> section below.</p>
<p>All flags are tunable such that their parameters can be configured to a variety of use cases and sequencing methods.</p>
<section id="rationale-assumptions-limitations">
<h2>Rationale, Assumptions &amp; Limitations<a class="headerlink" href="#rationale-assumptions-limitations" title="Link to this heading"></a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This is important!</p>
</div>
<p><code class="docutils literal notranslate"><span class="pre">hairpin2</span></code> is broadly speaking a collection of tests used in associating a given variant call with secondary structure formation during amplification (ADF, DVF, ALF), poorly mapped reads (ALF), or low-quality reads as defined by the user (LQF). The core rationale is to allow variant filtering based on a set of heuristics not considered by the variant caller. These range from relatively complex considerations to equally important but more simple ones. An example of a complex consideration might be accounting for a variant caller which does not compensate for secondary structure formation via the ADF flag. An example of a more simple consideration might be as follows; perhaps a user knows that, for a their sequencing type, reads with a mapping quality below X should not be considered for mutation calls, but this is unable to be expressed to the variant caller - spurious mutations from this issue can be accounted for via the LQF flag. In all cases, the appropriate parameters are determined by sequencing type, sequencing conditions, choice of aligner, and choice of variant caller. It is worth keeping this framing in mind during configuration and usage.</p>
<p>The hairpin2 tests have been designed with SNVs and small (2-3bp) indels in mind. If a mutation does not fall into that category, then no guarantee is presently made about the validity of the result. This is not to say that the tests certainly don’t apply, only that it is an unsupported use case, and should be approached with caution. The same is also true of sequencing type; the basis of the ADF and DVF tests in particular were formed explicitly around short-read amplification-based sequencing technologies. Usage on other sequencing types is therefore also unsupported.</p>
<p>hairpin2 requires paired-end data where alignment records have the <code class="docutils literal notranslate"><span class="pre">MC</span></code> tag and the complete CIGAR string is present in the <code class="docutils literal notranslate"><span class="pre">CIGAR</span></code> field (rather than the <code class="docutils literal notranslate"><span class="pre">CG:B,I</span></code> tag). If the <code class="docutils literal notranslate"><span class="pre">MC</span></code> tag is not present in your data, it can be added using <code class="docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">fixmate</span></code> or <code class="docutils literal notranslate"><span class="pre">biobambam2</span> <span class="pre">bamsormadup</span></code>. The tool can handle substitions, insertions, and deletions formatted per the VCF specification. At this time, the tool will not investigate mutations notated with angle brackets, e.g. <code class="docutils literal notranslate"><span class="pre">&lt;DEL&gt;</span></code>, complex mutations, or monomorphic reference. No further assumptions are made – other alignment tags and VCF fields are used, however they are mandatory per the relevant format specifications.</p>
</section>
<section id="processes-and-parameters">
<span id="process-target"></span><h2>Processes and Parameters<a class="headerlink" href="#processes-and-parameters" title="Link to this heading"></a></h2>
<p>For an input VCF, <code class="docutils literal notranslate"><span class="pre">hairpin2</span></code> will iterate over the records therein and analyse each record via a series of interdependent scientific steps.</p>
<p>This section details each process and its relevant parameters (if any) and outcomes, in execution order. The connection/interdependence between steps is also described. The parameters are also shown in their relevant TOML table as written in a hairpin2 TOML config at the top of each section.</p>
<p>For each variant examined, hairpin2 determines the mutation type (SUB, INS, DEL), fetches all reads covering the mutant position from the alignments, and then walks through the steps described in this section. A reference table is provided below for a high level overview to refer back to while reading about each process in more detail.</p>
<section id="process-relationships-reference">
<h3>Process Relationships Reference<a class="headerlink" href="#process-relationships-reference" title="Link to this heading"></a></h3>
<table class="docutils align-default" id="id1">
<caption><span class="caption-text"><strong>Read Taggers</strong></span><a class="headerlink" href="#id1" title="Link to this table"></a></caption>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Tests Reads With</p></th>
<th class="head"><p>Excludes Reads With</p></th>
<th class="head"><p>Adds Tag</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>mark-support</p></td>
<td><p>Any</p></td>
<td><p>None</p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p>mark-overlap</p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a></p></td>
<td><p>None</p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.OVERLAP_TAG" title="hairpin2.const.Tags.OVERLAP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">OVERLAP_TAG</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p>mark-low-qual</p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a></p></td>
<td><p>None</p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.LOW_QUAL_TAG" title="hairpin2.const.Tags.LOW_QUAL_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">LOW_QUAL_TAG</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p>mark-duplicates</p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a></p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.LOW_QUAL_TAG" title="hairpin2.const.Tags.LOW_QUAL_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">LOW_QUAL_TAG</span></code></a></p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.STUTTER_DUP_TAG" title="hairpin2.const.Tags.STUTTER_DUP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">STUTTER_DUP_TAG</span></code></a></p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default" id="id2">
<caption><span class="caption-text"><strong>Variant Flaggers</strong></span><a class="headerlink" href="#id2" title="Link to this table"></a></caption>
<thead>
<tr class="row-odd"><th class="head"><p>Flag</p></th>
<th class="head"><p>Tests Reads With</p></th>
<th class="head"><p>Excludes Reads With</p></th>
<th class="head"><p>Examines Tag</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>LQF</p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a></p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.OVERLAP_TAG" title="hairpin2.const.Tags.OVERLAP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">OVERLAP_TAG</span></code></a></p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.LOW_QUAL_TAG" title="hairpin2.const.Tags.LOW_QUAL_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">LOW_QUAL_TAG</span></code></a>, <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.STUTTER_DUP_TAG" title="hairpin2.const.Tags.STUTTER_DUP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">STUTTER_DUP_TAG</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p>DVF</p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a></p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.OVERLAP_TAG" title="hairpin2.const.Tags.OVERLAP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">OVERLAP_TAG</span></code></a>, <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.LOW_QUAL_TAG" title="hairpin2.const.Tags.LOW_QUAL_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">LOW_QUAL_TAG</span></code></a></p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.STUTTER_DUP_TAG" title="hairpin2.const.Tags.STUTTER_DUP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">STUTTER_DUP_TAG</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p>ALF</p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a></p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.OVERLAP_TAG" title="hairpin2.const.Tags.OVERLAP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">OVERLAP_TAG</span></code></a>, <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.LOW_QUAL_TAG" title="hairpin2.const.Tags.LOW_QUAL_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">LOW_QUAL_TAG</span></code></a>, <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.STUTTER_DUP_TAG" title="hairpin2.const.Tags.STUTTER_DUP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">STUTTER_DUP_TAG</span></code></a></p></td>
<td><p>None</p></td>
</tr>
<tr class="row-odd"><td><p>ADF</p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a></p></td>
<td><p><a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.OVERLAP_TAG" title="hairpin2.const.Tags.OVERLAP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">OVERLAP_TAG</span></code></a>, <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.LOW_QUAL_TAG" title="hairpin2.const.Tags.LOW_QUAL_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">LOW_QUAL_TAG</span></code></a>, <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.STUTTER_DUP_TAG" title="hairpin2.const.Tags.STUTTER_DUP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">STUTTER_DUP_TAG</span></code></a></p></td>
<td><p>None</p></td>
</tr>
</tbody>
</table>
</section>
<section id="read-taggers">
<h3>Read Taggers<a class="headerlink" href="#read-taggers" title="Link to this heading"></a></h3>
<hr class="docutils" />
<p>These processes execute prior to flaggers, examining the reads which cover the variant position. They apply extended tags to them such that downstream processes can filter/exclude/include reads from their analyses as needed. Note that tags applied to reads are internal, and are not written back to the alignment file.</p>
<section id="mark-support">
<h4>mark-support<a class="headerlink" href="#mark-support" title="Link to this heading"></a></h4>
<p><code class="docutils literal notranslate"><span class="pre">mark-support</span></code> takes no parameters. This process uses <a class="reference internal" href="hairpin2.sci_funcs.html#hairpin2.sci_funcs.TagSupportingReads" title="hairpin2.sci_funcs.TagSupportingReads"><code class="xref py py-class docutils literal notranslate"><span class="pre">TagSupportingReads</span></code></a> to tag supporting reads for the variant in question. In short the process checks that a given read:
- has all appropriate data/fields
- is not excluded by flag 0xE00
- that the alignment at the variant position matches the mutation call, both in terms of operation (INS/DEL/MATCH) and by displayed nucleotides</p>
<p>Reads found to support the variant are marked with the tag <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a></p>
<p>mark-support has no dependence on other steps</p>
</section>
<section id="mark-overlap">
<h4>mark-overlap<a class="headerlink" href="#mark-overlap" title="Link to this heading"></a></h4>
<p><code class="docutils literal notranslate"><span class="pre">mark-overlap</span></code> takes no parameters. This process uses <code class="xref py py-class docutils literal notranslate"><span class="pre">TagFragmentOverlapReads</span></code> to tag overlapping members of fragments (read pairs).</p>
<p>The second read found for a given qname is marked with the tag <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.OVERLAP_TAG" title="hairpin2.const.Tags.OVERLAP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">OVERLAP_TAG</span></code></a></p>
<p>mark-overlap only operates on reads marked with <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a> by <a class="reference internal" href="#mark-support"><span class="xref myst">mark-support</span></a> - i.e. those that supporting the variant.</p>
</section>
<section id="mark-low-qual">
<h4>mark-low-qual<a class="headerlink" href="#mark-low-qual" title="Link to this heading"></a></h4>
<div class="highlight-toml notranslate"><div class="highlight"><pre><span></span><span class="k">[params.mark-low-qual]</span>
<span class="n">min_avg_clip_quality</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">35</span>
<span class="n">min_mapping_quality</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">11</span>
<span class="n">min_base_quality</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">25</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">mark-low-qual</span></code> uses <a class="reference internal" href="hairpin2.sci_funcs.html#hairpin2.sci_funcs.TagLowQualReads" title="hairpin2.sci_funcs.TagLowQualReads"><code class="xref py py-class docutils literal notranslate"><span class="pre">TagLowQualReads</span></code></a> to mark reads that fail quality assessment.</p>
<p>Reads found to be of insufficient quality are marked with the tag <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.LOW_QUAL_TAG" title="hairpin2.const.Tags.LOW_QUAL_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">LOW_QUAL_TAG</span></code></a>. This step is essential to the LQF flag.</p>
<p>mark-low-qual only operates on reads marked with <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a> by <a class="reference internal" href="#mark-support"><span class="xref myst">mark-support</span></a>, i.e. those that support the variant.</p>
<table class="docutils align-default" id="id3" style="width: 100">
<caption><span class="caption-text"><strong>Parameters</strong></span><a class="headerlink" href="#id3" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>min_avg_clip_quality</p></td>
<td><p>Minimum mean quality of aligned bases for a read which contains soft clipping.</p></td>
</tr>
<tr class="row-odd"><td><p>min_mapping_quality</p></td>
<td><p>Minimum mapping quality (aka mapq) for a read.</p></td>
</tr>
<tr class="row-even"><td><p>min_base_quality</p></td>
<td><p>Minimum base quality for any position covering the variant in a read.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="mark-duplicates">
<h4>mark-duplicates<a class="headerlink" href="#mark-duplicates" title="Link to this heading"></a></h4>
<div class="highlight-toml notranslate"><div class="highlight"><pre><span></span><span class="k">[params.mark-duplicates]</span>
<span class="n">duplication_window_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">6</span>
</pre></div>
</div>
<div class="admonition-hidden-stutter-duplicates admonition" id="dup-explain-target">
<p class="admonition-title">Hidden Stutter Duplicates</p>
<p>In regions of low complexity, homopolymer tracts and secondary structure can cause PCR stuttering. PCR stuttering can lead to, for example, an erroneous additional A on the read when amplifying a tract of As. If duplicated reads contain stutter, this can lead to variation of read length and alignment to reference between reads that are in fact duplicates. Because of this variation, these duplicates both evade dupmarking and give rise to spurious variants when calling.</p>
</div>
<p><code class="docutils literal notranslate"><span class="pre">mark-duplicates</span></code> uses <a class="reference internal" href="hairpin2.sci_funcs.html#hairpin2.sci_funcs.TagStutterDuplicateReads" title="hairpin2.sci_funcs.TagStutterDuplicateReads"><code class="xref py py-class docutils literal notranslate"><span class="pre">TagStutterDuplicateReads</span></code></a> to mark reads which appear to be hidden PCR duplicates.
The algorithm is simply based on assessing the variation of endpoints amongst the tested reads. If a read start/end and it’s mate start/end
are within <code class="docutils literal notranslate"><span class="pre">duplication_window_size</span></code> of the closest neighbouring read’s start/end and mate start/end then those reads are considered to
be duplicates of each other.</p>
<p>Reads found to be duplicates are marked with the tag <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.STUTTER_DUP_TAG" title="hairpin2.const.Tags.STUTTER_DUP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">STUTTER_DUP_TAG</span></code></a>, excepting the highest quality read of the suspected duplicates. This step is essential to the DVF flag.</p>
<p>mark-duplicates only operates on reads marked with <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a> by <a class="reference internal" href="#mark-support"><span class="xref myst">mark-support</span></a>, i.e. those that support the variant.</p>
<table class="docutils align-default" id="id4" style="width: 100">
<caption><span class="caption-text"><strong>Parameters</strong></span><a class="headerlink" href="#id4" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>duplication_window_size</p></td>
<td><p>Alignment start/end window within which groups of fragments may be considered to be duplicates of one another.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="variant-flaggers">
<h3>Variant Flaggers<a class="headerlink" href="#variant-flaggers" title="Link to this heading"></a></h3>
<hr class="docutils" />
<p>The following processes examine the reads covering the variant position, after they have been appropriately tagged by the read tagging processes described above. Beyond simpy flagging the variant in the FILTER field, each flagger process details the outcome of its test directly into the INFO field of the VCF record. Therefore in addition to a description of the process and relevant parameters, a description of the possible outcome reasoning is also provided. Note that any INFO field set by hairpin2 can be expanded into a more human-readable description via the <a class="reference internal" href="quickstart.html#explain-qs-target"><span class="std std-ref">explain-var</span></a> subtool. For all cases, the high-level outcomes are as follows:</p>
<table class="docutils align-default" id="id5">
<caption><span class="caption-text"><strong>Outcomes</strong></span><a class="headerlink" href="#id5" title="Link to this table"></a></caption>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>PASS</p></td>
<td><p>The variant passed the test, and therefore was not flagged.</p></td>
</tr>
<tr class="row-odd"><td><p>FAIL</p></td>
<td><p>The variant failed the test, and therefore was flagged.</p></td>
</tr>
<tr class="row-even"><td><p>NA</p></td>
<td><p>The test could not be carried out.</p></td>
</tr>
</tbody>
</table>
<section id="lqf">
<h4>LQF<a class="headerlink" href="#lqf" title="Link to this heading"></a></h4>
<div class="highlight-toml notranslate"><div class="highlight"><pre><span></span><span class="k">[params.LQF]</span>
<span class="n">read_loss_threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">0.99</span>
<span class="n">min_pass_reads</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">2</span>
<span class="n">nsamples_threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">0</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">LQF</span></code> uses <a class="reference internal" href="hairpin2.sci_funcs.html#hairpin2.sci_funcs.ProportionBasedTest" title="hairpin2.sci_funcs.ProportionBasedTest"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProportionBasedTest</span></code></a> to flag variants which are largely supported by low quality reads.</p>
<p>The test is performed by examining the proportion of reads supporting a variant marked with <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.STUTTER_DUP_TAG" title="hairpin2.const.Tags.STUTTER_DUP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">STUTTER_DUP_TAG</span></code></a> by <a class="reference internal" href="#mark-duplicates"><span class="xref myst">mark-duplicates</span></a> and <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.LOW_QUAL_TAG" title="hairpin2.const.Tags.LOW_QUAL_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">LOW_QUAL_TAG</span></code></a> by <a class="reference internal" href="#mark-low-qual"><span class="xref myst">mark-low-qual</span></a>, compared to the total number of supporting reads (i.e. those tagged with <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a> by <a class="reference internal" href="#mark-support"><span class="xref myst">mark-support</span></a>).</p>
<p>LQF excludes from testing any reads tagged with <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.OVERLAP_TAG" title="hairpin2.const.Tags.OVERLAP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">OVERLAP_TAG</span></code></a> by <a class="reference internal" href="#mark-overlap"><span class="xref myst">mark-overlap</span></a> to avoid double counting.</p>
<table class="docutils align-default" id="id6" style="width: 100">
<caption><span class="caption-text"><strong>Parameters</strong></span><a class="headerlink" href="#id6" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p><strong>Name</strong></p></th>
<th class="head"><p><strong>Description</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>read_loss_threshold</p></td>
<td><p>Threshold of low-quality reads against total reads. If this threshold is exceeded, there are too many low-quality reads, and the variant is flagged.</p></td>
</tr>
<tr class="row-odd"><td><p>min_pass_reads</p></td>
<td><p>Minimum number of supporting reads without <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.STUTTER_DUP_TAG" title="hairpin2.const.Tags.STUTTER_DUP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">STUTTER_DUP_TAG</span></code></a> and <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.LOW_QUAL_TAG" title="hairpin2.const.Tags.LOW_QUAL_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">LOW_QUAL_TAG</span></code></a> necessary for a variant to pass the test. If there are fewer, the variant is flagged.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default" id="id7" style="width: 100">
<caption><span class="caption-text"><strong>INFO Conditions</strong></span><a class="headerlink" href="#id7" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>NO_READS</p></td>
<td><p>Check for suffcient reads available for testing, after excludes.</p></td>
</tr>
<tr class="row-odd"><td><p>THRESHOLD</p></td>
<td><p>Check for proportion of reads with stutter or low quality tags against the provided threshold.</p></td>
</tr>
<tr class="row-even"><td><p>MIN_PASS</p></td>
<td><p>Check for sufficient reads without stutter or low quality tags against the provided minimum.</p></td>
</tr>
</tbody>
</table>
<p>In the case of a LQF <code class="docutils literal notranslate"><span class="pre">PASS</span></code>, all three condiitions will be noted in the INFO field, since the variant must necessarily have passed all three. A <code class="docutils literal notranslate"><span class="pre">FAIL</span></code> may have <code class="docutils literal notranslate"><span class="pre">THRESHOLD</span></code>, <code class="docutils literal notranslate"><span class="pre">MIN_PASS</span></code>, or both. <code class="docutils literal notranslate"><span class="pre">NA</span></code> will only be seen with <code class="docutils literal notranslate"><span class="pre">NO_READS</span></code>. The INFO field for LQF additionally contains the proportion of reads tested found to have the <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.LOW_QUAL_TAG" title="hairpin2.const.Tags.LOW_QUAL_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">LOW_QUAL_TAG</span></code></a> or <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.STUTTER_DUP_TAG" title="hairpin2.const.Tags.STUTTER_DUP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">STUTTER_DUP_TAG</span></code></a>.</p>
</section>
<section id="dvf">
<h4>DVF<a class="headerlink" href="#dvf" title="Link to this heading"></a></h4>
<div class="highlight-toml notranslate"><div class="highlight"><pre><span></span><span class="k">[params.DVF]</span>
<span class="n">read_loss_threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">0.49</span>
<span class="n">min_pass_reads</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">2</span>
<span class="n">nsamples_threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">0</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">DVF</span></code> uses <a class="reference internal" href="hairpin2.sci_funcs.html#hairpin2.sci_funcs.ProportionBasedTest" title="hairpin2.sci_funcs.ProportionBasedTest"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProportionBasedTest</span></code></a> to flag variants which are largely supported by reads which appear to be PCR stutter duplicates as determined by the <a class="reference internal" href="#mark-duplicates"><span class="xref myst">mark-duplicates</span></a> process.</p>
<p>The test is performed by examining the proportion of reads supporting a variant marked with <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.STUTTER_DUP_TAG" title="hairpin2.const.Tags.STUTTER_DUP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">STUTTER_DUP_TAG</span></code></a> compared to the total number of supporting reads (i.e. those tagged with <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a>). Note that this does not consider reads marked as duplicates in the alignment file itself, as it is assumed that such duplication will have been addressed by the variant caller in making the call. The aim here is to find otherwise hidden duplicates that a variant caller is unlikely to have accounted for. Also note that since the parameters are independent from <a class="reference internal" href="hairpin2.process_wrappers.LQF.html#module-hairpin2.process_wrappers.LQF" title="hairpin2.process_wrappers.LQF"><span class="xref myst py py-mod">LQF</span></a>, the results of these two flags do not simply overlap.</p>
<p>DVF excludes from testing any reads tagged with <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.OVERLAP_TAG" title="hairpin2.const.Tags.OVERLAP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">OVERLAP_TAG</span></code></a> by <a class="reference internal" href="#mark-overlap"><span class="xref myst">mark-overlap</span></a> to avoid double counting, or <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.LOW_QUAL_TAG" title="hairpin2.const.Tags.LOW_QUAL_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">LOW_QUAL_TAG</span></code></a> by <a class="reference internal" href="#mark-low-qual"><span class="xref myst">mark-low-qual</span></a>.</p>
<table class="docutils align-default" id="id8" style="width: 100">
<caption><span class="caption-text"><strong>Parameters</strong></span><a class="headerlink" href="#id8" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>read_loss_threshold</p></td>
<td><p>threshold of duplicate reads against total reads.
If this threshold is exceeded, there are too many duplicate reads,
and the variant is flagged.</p></td>
</tr>
<tr class="row-odd"><td><p>min_pass_reads</p></td>
<td><p>minimum number of supporting reads without <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.STUTTER_DUP_TAG" title="hairpin2.const.Tags.STUTTER_DUP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">STUTTER_DUP_TAG</span></code></a> necessary for a variant to pass the test. If there are fewer, the variant is flagged.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default" id="id9" style="width: 100">
<caption><span class="caption-text"><strong>INFO Conditions</strong></span><a class="headerlink" href="#id9" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>NO_READS</p></td>
<td><p>Check for suffcient reads available for testing, after excludes.</p></td>
</tr>
<tr class="row-odd"><td><p>THRESHOLD</p></td>
<td><p>Check for proportion of reads with the stutter tag against the provided threshold.</p></td>
</tr>
<tr class="row-even"><td><p>MIN_PASS</p></td>
<td><p>Check for sufficient reads without the stutter tag against the provided minimum.</p></td>
</tr>
</tbody>
</table>
<p>In the case of a DVF <code class="docutils literal notranslate"><span class="pre">PASS</span></code>, all three condiitions will be noted in the INFO field, since the variant must necessarily have passed all three. A <code class="docutils literal notranslate"><span class="pre">FAIL</span></code> may have <code class="docutils literal notranslate"><span class="pre">THRESHOLD</span></code>, <code class="docutils literal notranslate"><span class="pre">MIN_PASS</span></code>, or both. <code class="docutils literal notranslate"><span class="pre">NA</span></code> will only be seen with <code class="docutils literal notranslate"><span class="pre">NO_READS</span></code>. The INFO field for DVF additionally contains the proportion of reads tested found to have the <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.STUTTER_DUP_TAG" title="hairpin2.const.Tags.STUTTER_DUP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">STUTTER_DUP_TAG</span></code></a>.</p>
</section>
<section id="alf">
<h4>ALF<a class="headerlink" href="#alf" title="Link to this heading"></a></h4>
<div class="highlight-toml notranslate"><div class="highlight"><pre><span></span><span class="k">[params.ALF]</span>
<span class="n">avg_AS_threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">0.93</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">ALF</span></code> uses <a class="reference internal" href="hairpin2.sci_funcs.html#hairpin2.sci_funcs.AlignmentScoreTest" title="hairpin2.sci_funcs.AlignmentScoreTest"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignmentScoreTest</span></code></a> to flag variants where the mean alignment score per base is less than <code class="docutils literal notranslate"><span class="pre">avg_AS_threshold</span></code>. The calcuation for each read is simply the value retrieved from the <code class="docutils literal notranslate"><span class="pre">AS</span></code> SAM tag divided by the read length. Users familiar with previous incarnations of LCMB variant post-processing should note that this is largely equivalent to <code class="docutils literal notranslate"><span class="pre">ASRD</span></code>.</p>
<p>ALF operates only on reads tagged with <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a> by <a class="reference internal" href="#mark-support"><span class="xref myst">mark-support</span></a>, i.e. only reads that support the given variant.</p>
<p>ALF excludes from testing any reads tagged with <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.LOW_QUAL_TAG" title="hairpin2.const.Tags.LOW_QUAL_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">LOW_QUAL_TAG</span></code></a> by <a class="reference internal" href="#mark-low-qual"><span class="xref myst">mark-low-qual</span></a>, <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.OVERLAP_TAG" title="hairpin2.const.Tags.OVERLAP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">OVERLAP_TAG</span></code></a> by <a class="reference internal" href="#mark-overlap"><span class="xref myst">mark-overlap</span></a>, or <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.STUTTER_DUP_TAG" title="hairpin2.const.Tags.STUTTER_DUP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">STUTTER_DUP_TAG</span></code></a> by <a class="reference internal" href="#mark-duplicates"><span class="xref myst">mark-duplicates</span></a>.</p>
<table class="docutils align-default" id="id10" style="width: 100">
<caption><span class="caption-text"><strong>Parameters</strong></span><a class="headerlink" href="#id10" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>avg_AS_threshold</p></td>
<td><p>threshold of mean alignment score per base. If this threshold is not reached, mean alignment score per base is too low, and the variant is flagged.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default" id="id11" style="width: 100">
<caption><span class="caption-text"><strong>INFO Conditions</strong></span><a class="headerlink" href="#id11" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>NO_READS</p></td>
<td><p>Check for suffcient reads available for testing, after excludes.</p></td>
</tr>
<tr class="row-odd"><td><p>INSUFFICIENT_AS_TAGS</p></td>
<td><p>Check for sufficient reads with AS tag to perform the test</p></td>
</tr>
<tr class="row-even"><td><p>ON_THRESHOLD</p></td>
<td><p>Check for sufficient reads remaining without stutter or low quality tags</p></td>
</tr>
</tbody>
</table>
<p>In the case of a ALF <code class="docutils literal notranslate"><span class="pre">PASS</span></code>, all three condiitions will be noted in the INFO field, since the variant must necessarily have passed all three. A <code class="docutils literal notranslate"><span class="pre">FAIL</span></code> will only have <code class="docutils literal notranslate"><span class="pre">ON_THRESHOLD</span></code>. <code class="docutils literal notranslate"><span class="pre">NA</span></code> will be seen with one of <code class="docutils literal notranslate"><span class="pre">NO_READS</span></code> or <code class="docutils literal notranslate"><span class="pre">INSUFFICIENT_AS_TAGS</span></code>. The INFO field for ALF additionally contains the mean alignment score per base seen on reads tested for the variant.</p>
</section>
<section id="adf">
<h4>ADF<a class="headerlink" href="#adf" title="Link to this heading"></a></h4>
<div class="highlight-toml notranslate"><div class="highlight"><pre><span></span><span class="k">[params.ADF]</span>
<span class="n">edge_definition</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">0.15</span>
<span class="n">edge_clustering_threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">0.9</span>
<span class="n">min_MAD_one_strand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">0.0</span>
<span class="n">min_sd_one_strand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">4.0</span>
<span class="n">min_MAD_both_strand_weak</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">2.0</span>
<span class="n">min_sd_both_strand_weak</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">2.0</span>
<span class="n">min_MAD_both_strand_strong</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">1.0</span>
<span class="n">min_sd_both_strand_strong</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">10.0</span>
<span class="n">low_n_supporting_reads_boundary</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">1</span>
<span class="n">min_non_edge_reads</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">0</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">ADF</span></code> uses <a class="reference internal" href="hairpin2.sci_funcs.html#hairpin2.sci_funcs.AnomalousDistributionTest" title="hairpin2.sci_funcs.AnomalousDistributionTest"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnomalousDistributionTest</span></code></a> to flag variants where the variant position is anomalously distributed on supporting reads. This often indicates artefacts due to secondary structure, but can also be the result of other artefactual phenomena.</p>
<p>ADF is an implementation of the following conditional described by <a class="reference external" href="https://www.nature.com/articles/s41596-020-00437-6">Ellis et al, 2020</a>. The full text of the conditional is reproduced below, with editorials in []. There is a point of ambiguity in the original conditional; the interpretation that this tool has opted for is indicated by [] and is expanded upon subsequently.</p>
<div class="quote admonition">
<p class="admonition-title">From Ellis et al.:</p>
<p>For each variant, if the number of variant-supporting reads determined is low (i.e. 0–1
reads) for one strand, follow Option 1. For each variant, if both strands have sufficient variant-supported reads
(i.e. ≥2 reads), follow Option 2.</p>
<ul class="simple">
<li><p>Option A - Low number of variant-supporting reads on one strand</p>
<ul>
<li><p>(i) For each variant, if one strand had too few variant-supporting reads, the other strand must
conform to:</p>
<ul>
<li><p>Fewer than 90% of variant-supporting reads [ON THE STRAND] have the variant located within the first 15% of the read measured from the alignment start position.</p></li>
<li><p>MAD &gt;0 [MEDIAN ABSOLUTE DEVIATION] and s.d. &gt;4 for that strand.</p></li>
</ul>
</li>
</ul>
</li>
<li><p>Option B - Sufficient variant-supporting reads on both strands</p>
<ul>
<li><p>(i) For each variant, if both strands have sufficient variant-supporting reads (i.e., ≥2 reads),
then one of the following must be true:</p>
<ul>
<li><p>Fewer than 90% of variant-supporting reads should have the variant located within the first 15% of the read measured from the alignment start position.</p></li>
<li><p>MAD &gt;2 and s.d. &gt;2 for both strands.</p></li>
<li><p>MAD &gt;1 and s.d. &gt;10 for one strand (i.e., strong evidence of high variability in variant position in variant-supporting reads).</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<p>The point of ambiguity is whether or not, on path A, to include the single read from the strand which does not sufficiently
support the variant in the test of positional distribution across the supporting reads.
The present interpretation is that the test should be applied only to the reads on the strand which has sufficient variant-supporting reads,
since the phrasing “the other strand must conform to” implies the exclusion of the single read on the low-support
strand.</p>
<p>ADF operates only on reads tagged with <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.SUPPORT_TAG" title="hairpin2.const.Tags.SUPPORT_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SUPPORT_TAG</span></code></a> by <a class="reference internal" href="#mark-support"><span class="xref myst">mark-support</span></a>, i.e. only reads that support the given variant.</p>
<p>ADF excludes from testing any reads tagged with <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.LOW_QUAL_TAG" title="hairpin2.const.Tags.LOW_QUAL_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">LOW_QUAL_TAG</span></code></a> by <a class="reference internal" href="#mark-low-qual"><span class="xref myst">mark-low-qual</span></a>, <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.OVERLAP_TAG" title="hairpin2.const.Tags.OVERLAP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">OVERLAP_TAG</span></code></a> by <a class="reference internal" href="#mark-overlap"><span class="xref myst">mark-overlap</span></a>, or <a class="reference internal" href="hairpin2.const.html#hairpin2.const.Tags.STUTTER_DUP_TAG" title="hairpin2.const.Tags.STUTTER_DUP_TAG"><code class="xref py py-attr docutils literal notranslate"><span class="pre">STUTTER_DUP_TAG</span></code></a> by <a class="reference internal" href="#mark-duplicates"><span class="xref myst">mark-duplicates</span></a>.</p>
<table class="docutils align-default" id="id12" style="width: 100">
<caption><span class="caption-text"><strong>Parameters</strong></span><a class="headerlink" href="#id12" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 40.0%" />
<col style="width: 60.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>edge_definition</p></td>
<td><p>percentage fraction of a read to be considered the edge when interpreting the conditional above, where 15% is used.</p></td>
</tr>
<tr class="row-odd"><td><p>edge_clustering_threshold</p></td>
<td><p>percentage threshold for the maximum fraction of reads that may express the variant at read edge as defined by <code class="docutils literal notranslate"><span class="pre">edge_defintion</span></code> when interpreting the conditional above, where 90% is used.</p></td>
</tr>
<tr class="row-even"><td><p>min_MAD_one_strand</p></td>
<td><p>threshold for Median Absolute Deviation to be used on path A of the conditional above, where 0 is used.</p></td>
</tr>
<tr class="row-odd"><td><p>min_sd_one_strand</p></td>
<td><p>threshold for standard deviation to be used on path A of the condtional above, where 2 is used.</p></td>
</tr>
<tr class="row-even"><td><p>min_MAD_both_strand_weak</p></td>
<td><p>threshold for Median Absolute Deviation to be used on path B, list option 2 of the condtional above, where 2 is used.</p></td>
</tr>
<tr class="row-odd"><td><p>min_sd_both_strand_weak</p></td>
<td><p>threshold for standard deviation to be used on Path B, list option 2 of the conditional above, where 2 is used.</p></td>
</tr>
<tr class="row-even"><td><p>min_MAD_both_strand_strong</p></td>
<td><p>threshold for Median Absolute Deviation to be used on path B, list option 3 of the condtional above, where 1 is used.</p></td>
</tr>
<tr class="row-odd"><td><p>min_sd_both_strand_strong</p></td>
<td><p>threshold for standard deviation to be used on Path B, list option 3 of the conditional above, where 10 is used.</p></td>
</tr>
<tr class="row-even"><td><p>low_n_supporting_reads_boundary</p></td>
<td><p>threshold for the minimum number of variant-supporting reads to be considered “low”, per the first sentence of the conditional above, where 1 is used.</p></td>
</tr>
<tr class="row-odd"><td><p>min_non_edge_reads</p></td>
<td><p>threshold; at least N supporting reads express variant away from the read edge as defined by edge_definition.</p></td>
</tr>
</tbody>
</table>
<p><code class="docutils literal notranslate"><span class="pre">min_non_edge_reads</span></code> is an addition/extension from hairpin2 beyond the original test described by Ellis et al. In the default parameter configs provided, all values are set to those described in the original conditional (per the param block at the top of this section) - <code class="docutils literal notranslate"><span class="pre">min_non_edge_reads</span></code> is set to 0, effectively disabling it.</p>
<table class="docutils align-default" id="id13" style="width: 100">
<caption><span class="caption-text"><strong>INFO Conditions</strong></span><a class="headerlink" href="#id13" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>NO_READS</p></td>
<td><p>Check for reads available for testing, after excludes.</p></td>
</tr>
<tr class="row-odd"><td><p>INSUFFICIENT_READS</p></td>
<td><p>Check for minimum reads available for testing on at least one strand, as defined by <code class="docutils literal notranslate"><span class="pre">low_n_supporting_reads_boundary</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p>EDGE_CLUSTERING</p></td>
<td><p>Check for fewer than <code class="docutils literal notranslate"><span class="pre">edge_clustering_threshold</span></code>% of reads expressing variant within <code class="docutils literal notranslate"><span class="pre">edge_definition</span></code> of alignment start, from both Option A and B in Ellis et al.</p></td>
</tr>
<tr class="row-odd"><td><p>ONE_STRAND_DISTRIB</p></td>
<td><p>Check for MAD and standard deviation condition from Option A in Ellis et. al., as defined by <code class="docutils literal notranslate"><span class="pre">min_MAD_one_strand</span></code> and <code class="docutils literal notranslate"><span class="pre">min_sd_one_strand</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p>BOTH_STRAND_DISTRIB_BOTH</p></td>
<td><p>Check for MAD and standard deviation condition from Option B as applied to both strands in Ellis et al., as defined by <code class="docutils literal notranslate"><span class="pre">min_MAD_both_strand_weak</span></code> and <code class="docutils literal notranslate"><span class="pre">min_sd_both_strand_weak</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p>BOTH_STRAND_DISTRIB_ONE</p></td>
<td><p>Check for MAD and standard devation condition from Option B as applied to one strand in Ellis et al., as defined by <code class="docutils literal notranslate"><span class="pre">min_MAD_both_strand_strong</span></code> and <code class="docutils literal notranslate"><span class="pre">min_sd_both_strand_strong</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p>MIN_NON_EDGE</p></td>
<td><p>Check for suffcient reads expressing variant away from read edge, as defined by <code class="docutils literal notranslate"><span class="pre">edge</span> <span class="pre">definition</span></code>.</p></td>
</tr>
</tbody>
</table>
<p>In the case of an ADF <code class="docutils literal notranslate"><span class="pre">PASS</span></code>/<code class="docutils literal notranslate"><span class="pre">FAIL</span></code>, all associated conditions passed/failed will be found in the INFO field. The specific conditions therein depend on which branch of the conditional was used. The INFO field contains the strand/s to help indicate the branch taken. <code class="docutils literal notranslate"><span class="pre">NA</span></code> will be seen with either NO_READS or INSUFFICIENT_READS only.</p>
</section>
</section>
</section>
<section id="writing-a-config">
<h2>Writing a Config<a class="headerlink" href="#writing-a-config" title="Link to this heading"></a></h2>
<p>hairpin2 configs can be written in either JSON or TOML format. We recommend <a class="reference external" href="https://toml.io/en/">TOML</a>, but both are fully supported. In either case, for standard usage a config needs a single top level key, <code class="docutils literal notranslate"><span class="pre">params</span></code>, which will then contain keys for each process by title, e.g. <code class="docutils literal notranslate"><span class="pre">mark-duplicates</span></code>, <code class="docutils literal notranslate"><span class="pre">ADF</span></code>, etc. Finally, under the process keys come the parameters. This format is pleasant to express in TOML:</p>
<div class="highlight-toml notranslate"><div class="highlight"><pre><span></span><span class="k">[params.mark-duplicates]</span>
<span class="n">duplication_window_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">6</span>


<span class="k">[params.DVF]</span>
<span class="n">read_loss_threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mf">0.49</span>
<span class="n">min_pass_reads</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">2</span>
<span class="n">nsamples_threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">0</span>

<span class="c1"># other processes... (n.b. toml allows comments)</span>
</pre></div>
</div>
<p>if you prefer JSON:</p>
<div class="highlight-json notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
<span class="w">  </span><span class="nt">&quot;params&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nt">&quot;mark-low-qual&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">      </span><span class="nt">&quot;min_avg_clip_quality&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">35</span><span class="p">,</span>
<span class="w">      </span><span class="nt">&quot;min_mapping_quality&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">11</span><span class="p">,</span>
<span class="w">      </span><span class="nt">&quot;min_base_quality&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">25</span>
<span class="w">    </span><span class="p">},</span>
<span class="w">    </span><span class="nt">&quot;LQF&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">      </span><span class="nt">&quot;read_loss_threshold&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">0.99</span><span class="p">,</span>
<span class="w">      </span><span class="nt">&quot;min_pass_reads&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">2</span><span class="p">,</span>
<span class="w">      </span><span class="nt">&quot;nsamples_threshold&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">0</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">}</span>
</pre></div>
</div>
<p>LCMB default parameters are provided in both JSON and TOML format in the <code class="docutils literal notranslate"><span class="pre">example-configs/</span></code> directory found in the hairpin2 <a class="reference external" href="https://github.com/cancerit/hairpin2">GitHub repository</a>.</p>
</section>
<section id="understanding-decisions">
<span id="understanding-decisions-target"></span><h2>Understanding Decisions<a class="headerlink" href="#understanding-decisions" title="Link to this heading"></a></h2>
<p>hairpin2 stores detailed reasoning for each flagging decision in the INFO field of each tested variant. There will be VCF-spec standard key=value pairs for each flag tested. The key is the flag name, as would be applied to the FILTER column if the variant fails, e.g., <code class="docutils literal notranslate"><span class="pre">LQF</span></code>. For each flag, there will be as many key=value pairs as there are alts for the variant.</p>
<p>The value side of the key=value pair is formatted as pipe (<code class="docutils literal notranslate"><span class="pre">|</span></code>) separated fields. The standard fields shared by all flags are as follows:</p>
<table class="docutils align-default" id="id14" style="width: 100">
<caption><span class="caption-text"><strong>Value Fields</strong></span><a class="headerlink" href="#id14" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>alt</p></td>
<td><p>The alt to which this data applies</p></td>
</tr>
<tr class="row-even"><td><p>outcome</p></td>
<td><p>The result of testing for this alt (PASS, FAIL, NA)</p></td>
</tr>
<tr class="row-odd"><td><p>conditions</p></td>
<td><p>The conditions upon which the outcome is based, stored as a bitwise hex flag</p></td>
</tr>
<tr class="row-even"><td><p>nreads</p></td>
<td><p>The number of reads examined by this test, after filtering to appropriately tagged reads</p></td>
</tr>
</tbody>
</table>
<p>A flag may add any number of subsequent extra fields. So in full, a hairpin2 INFO key=value pair looks like <code class="docutils literal notranslate"><span class="pre">&lt;flag_name&gt;=&lt;alt&gt;|&lt;outcome&gt;|&lt;conditions&gt;|&lt;nreads&gt;|...</span></code>.</p>
<p>key=value paris for any hairpin2 flag can be input to the <a class="reference internal" href="quickstart.html#explain-qs-target"><span class="std std-ref">explain-var</span></a> tool to convert this data into a more human-readable format, including descriptions of any additional fields. All fields are also fully described in the VCF header.</p>
</section>
<section id="reproducibility-parameter-distribution">
<span id="repro-target"></span><h2>Reproducibility &amp; Parameter Distribution<a class="headerlink" href="#reproducibility-parameter-distribution" title="Link to this heading"></a></h2>
<p>hairpin2 uses only deterministic processes and so is reproducible within the standard computational bounds of that term. Singuarity and Docker definition files are provided for containerisation if needed.</p>
<p>hairpin2 stores the complete information necessary to distribute and reproduce a run directly in the VCF header. The following header keys are used</p>
<table class="docutils align-default" id="id15" style="width: 100">
<caption><span class="caption-text"><strong>Header Keys</strong></span><a class="headerlink" href="#id15" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>hairpin2_version</p></td>
<td><p>stores version of hairpin2 used</p></td>
</tr>
<tr class="row-even"><td><p>hairpin2_params</p></td>
<td><p>stores paramters and execution configuration</p></td>
</tr>
<tr class="row-odd"><td><p>hairpin2_samples</p></td>
<td><p>stores which samples in the VCF were selected for testing</p></td>
</tr>
</tbody>
</table>
<p>The parameters are stored in a reduced represenation and can be extracted from a VCF using <a class="reference internal" href="quickstart.html#get-params-qs-target"><span class="std std-ref">get-params</span></a> back into standard JSON format. It is not necessary to understand the way in which the parameters are encoded into the header (explained below) to use this functionality.</p>
<p>The hairpin2_params key stores data in a JSON compressed with <code class="docutils literal notranslate"><span class="pre">zlib</span></code> and encoded to an ascii string using <code class="docutils literal notranslate"><span class="pre">base85</span></code>. The result is a string of ascii characters, much shorter than the original JSON, which does not invite or allow accidental editing of the parameters once written into the header (either by hand or by another tool). Since zlib compression includes an error-checking checksum, the string is guaranteed to transform back into the exact parameters encoded. Base85 encoding ensures that only VCF-safe characters are used  per the VCF format spec.</p>
</section>
<section id="further-development">
<h2>Further Development<a class="headerlink" href="#further-development" title="Link to this heading"></a></h2>
<p>We are very open to making updates and improvements to the tool to support a wide variety of use cases. If you have a request, please get in touch via the <a class="reference external" href="https://github.com/cancerit/hairpin2">GitHub</a>.</p>
<!-- TODO!! -->
<!-- ## Troubleshooting -->
<!-- NA variants -->
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="quickstart.html" class="btn btn-neutral float-left" title="Quickstart" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="hairpin2.html" class="btn btn-neutral float-right" title="hairpin2 package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Alex Byrne.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>